### PR TITLE
Component Hot Reloading

### PR DESCRIPTION
**Pull request by JoshVanL**
Adds documentation for the preview feature `Component Hot Reloading`

Part of https://github.com/dapr/dapr/issues/1172

### CODE REVIEW

**Comment by hhunter-ms** [comment]
@JoshVanL could you also update this doc for hot reloading as well? https://docs.dapr.io/operations/components/component-updates/ - Suggestion is to simply add to this document that if you enable the hotloading feature then you get reloading (and still describe the current behaviour if not set)
 
- Add a section above 'Kubernetes' to briefly explain hotloading
- Reword anywhere that implies components can't be updated automatically
- Clarify if this applies to every component, including middleware?
- Will hot reloading be a default when not in preview?

File: daprdocs/content/en/concepts/components-concept.md [hunk 1]
```diff
@@ -52,6 +52,14 @@ For example:
 
 For more information read [Pluggable components overview]({{< ref "pluggable-components-overview" >}})
 
+## Hot Reloading
+
+With the [`HotReload` feature enabled]({{< ref "support-preview-features.md" >}}), Components are able to be "hot reloaded" at runtime.
```
Comment (msfussell) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
With the [`HotReload` feature enabled]({{< ref "support-preview-features.md" >}}), components are able to be "hot reloaded" at runtime.
```

File: daprdocs/content/en/concepts/components-concept.md [hunk 2]
```diff
@@ -52,6 +52,14 @@ For example:
...
+This means that you can update component configuration without restarting the Dapr runtime.
+Component reloading happens when a Component resource is created, updated, or deleted, either in the Kubernetes API or in Self Hosted mode when a file is changed in the `resources` directory.
```
Comment (msfussell) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
Component reloading happens when a component resource is created, updated, or deleted, either in the Kubernetes API or in Self Hosted mode when a file is changed in the `resources` directory.
```

File: daprdocs/content/en/concepts/components-concept.md [hunk 3]
```diff
@@ -52,6 +52,14 @@ For example:
...
+When a component is updated, the Component is first closed, and then reinitialized using the new configuration.
```
Comment (msfussell) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
When a component is updated, the component is first closed, and then reinitialized using the new configuration.
```

File: daprdocs/content/en/concepts/components-concept.md [hunk 4]
```diff
@@ -52,6 +52,14 @@ For example:
...
+There will be a short period of time where the component is unavailable during the duration of the reload until the component is reinitialized.
```
Comment (msfussell) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
There is a short period of time where the component is unavailable during the duration of the reload until the component is reinitialized.
```

**Comment by JoshVanL** [comment]
Thanks @hhunter-ms @msfussell, I have added your suggestions.

> Will hot reloading be a default when not in preview?

@hhunter-ms indeed, the plan to have this be the default and only option once HotReloading comes out of preview.

File: daprdocs/content/en/operations/components/component-updates.md [hunk 5]
```diff
@@ -6,18 +6,27 @@ weight: 300
 description: "Updating deployed components used by applications"
 ---
 
-When making an update to an existing deployed component used by an application, Dapr does not update the component automatically. The Dapr sidecar needs to be restarted in order to pick up the latest version of the component. How this done depends on the hosting environment.
+When making an update to an existing deployed component used by an application, Dapr does not update the component automatically unless the `HotReload` feature gate is enabled.
+The Dapr sidecar needs to be restarted in order to pick up the latest version of the component.
+How this done depends on the hosting environment.
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/components/component-updates.md]:
```suggestion
How this is done depends on the hosting environment.
```

File: daprdocs/content/en/operations/components/component-updates.md [hunk 6]
```diff
@@ -6,18 +6,27 @@ weight: 300
...
+This feature is currently in preview.
+{{% /alert %}}
 
 ## Kubernetes
 
 When running in Kubernetes, the process of updating a component involves two steps:
 
 1. Applying the new component YAML to the desired namespace
-2. Performing a [rollout restart operation](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#updating-resources) on your deployments to pick up the latest component
+2. (Unless the [`HotReload` feature gate is enabled]({{< ref "support-preview-features.md" >}})), perform a [rollout restart operation](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#updating-resources) on your deployments to pick up the latest component
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/components/component-updates.md]:
```suggestion
1. Unless the [`HotReload` feature gate is enabled]({{< ref "support-preview-features.md" >}}), perform a [rollout restart operation](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#updating-resources) on your deployments to pick up the latest component
```

File: daprdocs/content/en/operations/components/component-updates.md [hunk 7]
```diff
@@ -6,18 +6,27 @@ weight: 300
...
+This is enabled by via the [`HotReload` feature gate]({{< ref "support-preview-features.md" >}}).
+All component types are supported for hot reloading.
+This feature is currently in preview.
+{{% /alert %}}
 
 ## Kubernetes
 
 When running in Kubernetes, the process of updating a component involves two steps:
 
 1. Applying the new component YAML to the desired namespace
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/components/component-updates.md]:
```suggestion
1. Apply the new component YAML to the desired namespace
```

File: daprdocs/content/en/operations/components/component-updates.md [hunk 8]
```diff
@@ -6,18 +6,27 @@ weight: 300
...
-2. Performing a [rollout restart operation](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#updating-resources) on your deployments to pick up the latest component
+2. (Unless the [`HotReload` feature gate is enabled]({{< ref "support-preview-features.md" >}})), perform a [rollout restart operation](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#updating-resources) on your deployments to pick up the latest component
 
 ## Self Hosted
 
-When running in Self Hosted mode, the process of updating a component involves a single step of stopping the `daprd` process and starting it again to pick up the latest component.
+Unless the [`HotReload` feature gate is enabled]({{< ref "support-preview-features.md" >}}), the process of updating a component involves a single step of stopping the `daprd` process and starting it again to pick up the latest component.
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/components/component-updates.md]:
```suggestion
Unless the [`HotReload` feature gate is enabled]({{< ref "support-preview-features.md" >}}), the process of updating a component involves a single step of stopping and restarting the `daprd` process to pick up the latest component.
```

File: daprdocs/content/en/operations/support/support-preview-features.md [hunk 9]
```diff
@@ -22,4 +22,4 @@ For CLI there is no explicit opt-in, just the version that this was first made a
...
 | **Actor State TTL** | Allow actors to save records to state stores with Time To Live (TTL) set to automatically clean up old data. In its current implementation, actor state with TTL may not be reflected correctly by clients, read [Actor State Transactions]({{< ref actors_api.md >}}) for more information. | `ActorStateTTL` | [Actor State Transactions]({{< ref actors_api.md >}}) | v1.11  |
 | **Transactional Outbox** | Allows state operations for inserts and updates to be published to a configured pub/sub topic using a single transaction across the state store and the pub/sub | N/A | [Transactional Outbox Feature]({{< ref howto-outbox.md >}}) | v1.12  |
-
+| **Component Hot Reloading** | Allows for dapr loaded Components to be "hot reloaded". A Component spec will be reloaded when it is created/updated/deleted in Kubernetes or on file when running in Self-hosted mode.| `HotReload`| [Hot Reloading]({{< ref components-concept.md >}}) | v1.13  |
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/support/support-preview-features.md]:
```suggestion
| **Component Hot Reloading** | Allows for Dapr-loaded components to be "hot reloaded". A component spec is reloaded when it is created/updated/deleted in Kubernetes or on file when running in self-hosted mode.| `HotReload`| [Hot Reloading]({{< ref components-concept.md >}}) | v1.13  |
```

File: daprdocs/content/en/concepts/components-concept.md [hunk 10]
```diff
@@ -52,6 +52,14 @@ For example:
 
 For more information read [Pluggable components overview]({{< ref "pluggable-components-overview" >}})
 
+## Hot Reloading
+
+With the [`HotReload` feature enabled]({{< ref "support-preview-features.md" >}}), components are able to be "hot reloaded" at runtime.
+This means that you can update component configuration without restarting the Dapr runtime.
+Component reloading happens when a component resource is created, updated, or deleted, either in the Kubernetes API or in Self Hosted mode when a file is changed in the `resources` directory.
```
Comment (hhunter-ms) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
Component reloading occurs when a component resource is created, updated, or deleted, either in the Kubernetes API or in self-hosted mode when a file is changed in the `resources` directory.
```

File: daprdocs/content/en/concepts/components-concept.md [hunk 11]
```diff
@@ -52,6 +52,14 @@ For example:
...
+When a component is updated, the component is first closed, and then reinitialized using the new configuration.
+There is a short period of time where the component is unavailable during the duration of the reload until the component is reinitialized.
```
Comment (hhunter-ms) [inline daprdocs/content/en/concepts/components-concept.md]:
```suggestion
The component is unavailable for a short period of time during reload and reinitialization.
```

File: daprdocs/content/en/operations/components/component-updates.md [hunk 12]
```diff
@@ -6,18 +6,27 @@ weight: 300
 description: "Updating deployed components used by applications"
 ---
 
-When making an update to an existing deployed component used by an application, Dapr does not update the component automatically. The Dapr sidecar needs to be restarted in order to pick up the latest version of the component. How this done depends on the hosting environment.
+When making an update to an existing deployed component used by an application, Dapr does not update the component automatically unless the `HotReload` feature gate is enabled.
+The Dapr sidecar needs to be restarted in order to pick up the latest version of the component.
+How this done depends on the hosting environment.
+
+{{% alert title="Note" color="primary" %}}
+Dapr can be made to "hot reload" components where updates will be picked up automatically without needing a restart.
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/components/component-updates.md]:
```suggestion
Dapr can be made to "hot reload" components, where updates are picked up automatically without needing a restart.
```

**Review by hhunter-ms — CHANGES_REQUESTED** [review]
quick review

**Review by hhunter-ms — APPROVED** [review]

**Pull request merged**